### PR TITLE
Emulate suggestion mode in LSP completion by always soft-select

### PR DESCRIPTION
**Pull request by genlu**
Fix https://github.com/dotnet/vscode-csharp/issues/5732

### CODE REVIEW

File: src/Features/LanguageServer/Protocol/Handler/Completion/AbstractLspCompletionResultCreationService.cs [hunk 1]
```diff
@@ -132,18 +133,24 @@ async Task<LSP.CompletionItem> CreateLSPCompletionItemAsync(CompletionItem item,
...
-                    // datetime completion, etc. we create a completion item without any specific commit characters.
-                    // This means only tab / enter will commit. VS supports soft selection, so we only do this for non-VS clients.
-                    //
-                    // Note this only applies when user hasn't actually typed anything and completion provider does not request the item
-                    // to be hard-selected. Otherwise, we set its commit characters as normal. This also means we'd need to set IsIncomplete to true
+                    lspItem.CommitCharacters = GetCommitCharacters(item, commitCharactersRuleCache);
+                    return lspItem;
+                }
+
+                // VSCode does not have the concept of soft selection, the list is always hard selected.
```
Comment (davidwengier) [inline src/Features/LanguageServer/Protocol/Handler/Completion/AbstractLspCompletionResultCreationService.cs:142]: I'm curious about this. It's not just VSCode that doesn't have soft selection, VS doesn't have it either in the LSP client. Should the above block (checking `lspVSClientCapability`) even be present? Certainly we have this issue in VS in Razor, so is it just that nobody turns on LSP for C# files in VS, so this code is incorrect but we're just assuming it's correct?

**Review by arkalyanms — APPROVED** [review]

File: src/Features/LanguageServer/Protocol/Handler/Completion/AbstractLspCompletionResultCreationService.cs (hunk 1 above)
Comment (genlu) [reply src/Features/LanguageServer/Protocol/Handler/Completion/AbstractLspCompletionResultCreationService.cs:142]: VS client has suggestion mode (`VSInternalCompletionList.SuggestionMode`), so I assumed it has similar behavior s in VS. Maybe I was wrong, will need to try it.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/Completion/AbstractLspCompletionResultCreationService.cs:142]: > VS doesn't have it either in the LSP client.

Wat.

Comment (genlu) [reply src/Features/LanguageServer/Protocol/Handler/Completion/AbstractLspCompletionResultCreationService.cs:142]: This is the doc-comment for VS client's suggestion mode, which suggests there's soft-selection
```cs
 //
 // Summary:
 //     Gets or sets a value indicating whether the completion list should use suggestion
 //     mode. In suggestion mode items are "soft-selected" by default.
 [DataMember(Name = "_vs_suggestionMode")]
 [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
 public bool SuggestionMode { get; set; }
```

Comment (davidwengier) [reply src/Features/LanguageServer/Protocol/Handler/Completion/AbstractLspCompletionResultCreationService.cs:142]: TIL. Sounds like we need to try that! Thanks

**Comment by 333fred** [comment]
@genlu can you add a test case for typing `"` in an argument list? IE, `Console.WriteLine($$)`, typing `"` should insert just the `"`, not the top of the completion list. A user was running into this today with `editor.acceptSuggestionOnCommitCharacter` set to true, and I think this PR will fix that.

**Comment by genlu** [comment]
@333fred that should be fixed by [this PR](https://dnceng.visualstudio.com/internal/_git/dotnet-roslyn/pullrequest/31357). But I will try again to be sure

**Pull request merged**